kubernetes/controller/internal/configuration/config.go

```diff
@@ -0,0 +1,123 @@
+	"ocm.software/open-component-model/kubernetes/controller/api/v1alpha1"
+)
+
+// GetConfigFromSecret extracts and decodes OCM configuration from a Kubernetes Secret.
+// It looks for configuration data under the OCMConfigKey.
+func GetConfigFromSecret(secret *corev1.Secret) (*genericv1.Config, error) {
+	data, ok := secret.Data[v1alpha1.OCMConfigKey]
+	if !ok || len(data) == 0 {
+		return nil, nil
+	}
+
+	var cfg genericv1.Config
+	if err := genericv1.Scheme.Decode(bytes.NewReader(data), &cfg); err != nil {
+		return nil, fmt.Errorf("failed to decode ocm config from secret %s/%s: %w",
+			secret.Namespace, secret.Name, err)
+	}
+
+	return &cfg, nil
+}
+
+// GetConfigFromConfigMap extracts and decodes OCM configuration from a Kubernetes ConfigMap.
+// It looks for configuration data under the OCMConfigKey.
+func GetConfigFromConfigMap(configMap *corev1.ConfigMap) (*genericv1.Config, error) {
+	data, ok := configMap.Data[v1alpha1.OCMConfigKey]
+	if !ok || len(data) == 0 {
+		return nil, nil
+	}
+
+	var cfg genericv1.Config
+	if err := genericv1.Scheme.Decode(strings.NewReader(data), &cfg); err != nil {
+		return nil, fmt.Errorf("failed to decode ocm config from configmap %s/%s: %w",
+			configMap.Namespace, configMap.Name, err)
+	}
+
+	return &cfg, nil
+}
+
+// GetConfigFromObject extracts configuration from either a Secret or ConfigMap.
+func GetConfigFromObject(obj client.Object) (*genericv1.Config, error) {
+	switch o := obj.(type) {
+	case *corev1.Secret:
+		return GetConfigFromSecret(o)
+	case *corev1.ConfigMap:
+		return GetConfigFromConfigMap(o)
+	default:
+		return nil, fmt.Errorf("unsupported configuration object type: %T", obj)
+	}
+}
+
+// LoadConfigurations loads OCM configurations from a list of OCMConfiguration references.
+// It fetches the referenced Secrets/ConfigMaps from the cluster and extracts their configuration.
+// All configurations are merged using FlatMap into a single Config.
+func LoadConfigurations(ctx context.Context, k8sClient client.Reader, namespace string, ocmConfigs []v1alpha1.OCMConfiguration) (*genericv1.Config, error) {
+	if len(ocmConfigs) == 0 {
+		return &genericv1.Config{}, nil
+	}
+
+	// Best effort sort of the configs to ensure they are always processed in the same order.
+	sort.Slice(ocmConfigs, func(i, j int) bool {
+		key1 := ocmConfigs[i].Name + "." + ocmConfigs[i].Namespace
+		key2 := ocmConfigs[j].Name + "." + ocmConfigs[j].Namespace
+
+		return key1 < key2
+	})
+
+	configs := make([]*genericv1.Config, 0, len(ocmConfigs))
+	for _, ocmConfig := range ocmConfigs {
+		ns := ocmConfig.Namespace
+		if ns == "" {
+			ns = namespace
+		}
+
+		var obj client.Object
+		switch ocmConfig.Kind {
+		case "Secret":
+			obj = &corev1.Secret{}
+		case "ConfigMap":
+			obj = &corev1.ConfigMap{}
+		default:
+			return nil, fmt.Errorf("unsupported configuration kind: %s", ocmConfig.Kind)
+		}
+
+		key := client.ObjectKey{Namespace: ns, Name: ocmConfig.Name}
+		if err := k8sClient.Get(ctx, key, obj); err != nil {
```

Comment:
are we sure this cannot cause perf issues when accessing many objects?

Reply:
Hm, I could try and parallelise this.